### PR TITLE
[IMP] sale: down payment tax breakdown

### PR DESCRIPTION
**Pull request by jbw-odoo**
Break down payments per tax. When creating a down payment for a sale order, create one line per tax instead of one line for the entire order.

[Sale Order]
Product    Qty  Unit price    Tax   Subtotal
Product1   1    1000          10%   1000
Product2   1    1000          5%    1000
                                    ====
                            Untaxed 2000
                                Tax 150
                                    ====
                              Total 2150

Make a 50% (1075) down payment :

[BEFORE]
Product    Qty  Unit price    Tax   Subtotal
Product1   1    1000          10%   1000
Product2   1    1000          5%    1000
Down pay   1    1075                0
                                    ====
                            Untaxed 2000
                                Tax 150
                                    ====
                              Total 2150

[AFTER]
Product    Qty  Unit price    Tax   Subtotal
Product1   1    1000          10%   1000
Product2   1    1000          5%    1000
Down pay   1    500           10%   0
Down pay   1    500           5%    0
                                    ====
                            Untaxed 2000
                                Tax 150
                                    ====
                              Total 2150

Task: 2678531

Enterprise PR: https://github.com/odoo/enterprise/pull/36331

### CODE REVIEW

**Comment by jbw-odoo** [comment]
Prev PR https://github.com/odoo/odoo/pull/105177

**Comment by jbw-odoo** [comment]
Prev. PR https://github.com/odoo/odoo/pull/81337

File: addons/sale/wizard/sale_make_invoice_advance.py [hunk 1]
```diff
@@ -219,26 +219,82 @@ def _prepare_down_payment_section_values(self, order):
         del context
         return so_values
 
-    def _prepare_so_line_values(self, order):
+    def _prepare_so_lines_values(self, order):
+        self.ensure_one()
+        down_payment_amount = self._get_down_payment_amount(order)
+        group_by_taxes = defaultdict(lambda: {
```
Comment (smetl) [inline addons/sale/wizard/sale_make_invoice_advance.py:225]: you must never do such kind of computation yourself. Instead, use the helpers on `account.tax`. For example:
https://github.com/odoo/odoo/blob/16.0/addons/sale/models/sale_order_line.py#L598

This allows you to get the tax amounts (grouped or not depending the keys you want) for any business documents

Comment (jbw-odoo) [reply addons/sale/wizard/sale_make_invoice_advance.py:236]: Did that + simplified

File: addons/sale/wizard/sale_make_invoice_advance.py [hunk 2]
```diff
@@ -146,10 +147,9 @@ def _check_down_payment_product_is_valid(self):
     #=== ACTION METHODS ===#
 
     def create_invoices(self):
-        self._create_invoices(self.sale_order_ids)
-
+        invoices = self._create_invoices(self.sale_order_ids)
         if self.env.context.get('open_invoices'):
-            return self.sale_order_ids.action_view_invoice()
+            return self.sale_order_ids.action_view_invoice(invoices)
```
Comment (smetl) [inline addons/sale/wizard/sale_make_invoice_advance.py:152]:
```suggestion
            return self.sale_order_ids.action_view_invoice(invoices=invoices)
```

File: addons/sale/wizard/sale_make_invoice_advance.py (hunk 1 above)
Comment (smetl) [reply addons/sale/wizard/sale_make_invoice_advance.py:236]: I think you can simplify more using `_aggregate_taxes` by passing a custom `grouping_key_generator`

File: addons/sale/wizard/sale_make_invoice_advance.py (hunk 2 above)
Comment (jbw-odoo) [reply addons/sale/wizard/sale_make_invoice_advance.py]: > I think you can simplify more using _aggregate_taxes by passing a custom grouping_key_generator

I don't see how.

File: addons/sale/models/sale_order.py [hunk 3]
```diff
@@ -985,8 +991,9 @@ def _prepare_invoice(self):
             'invoice_line_ids': [],
         }
 
-    def action_view_invoice(self):
-        invoices = self.mapped('invoice_ids')
+    def action_view_invoice(self, invoices=False):
+        if not invoices:
+            invoices = self.mapped('invoice_ids')
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:996]: why is this change necessary ?
AFAIK, the invoices created by the wizard should be linked to the SOs anyway.

File: addons/sale/views/sale_order_views.xml [hunk 4]
```diff
@@ -226,11 +226,21 @@
...
+                    <button name="action_view_invoice"
+                        type="object"
+                        class="oe_stat_button text-danger"
+                        icon="fa-pencil-square-o"
+                        attrs="{'invisible': ['|', ('invoice_count', '=', 0), ('any_draft_invoice', '=', False)]}">
+                        <div class="o_field_widget o_stat_info">
+                            <field name="invoice_count" nolabel="1"/>
+                            <span class="o_stat_text">Invoices</span>
+                        </div>
+                    </button>
```
Comment (Feyensv) [inline addons/sale/views/sale_order_views.xml:238]: Is this really necessary only to show some text in red ?

File: addons/sale/models/sale_order.py (hunk 3 above)
Comment (jbw-odoo) [reply addons/sale/models/sale_order.py:996]: This is to (only) show the newly created down payment after the wizard. Imo i make sens to display the down payment instead of a list of the dowmnpayment + existing invoices

File: addons/sale/views/sale_order_views.xml (hunk 4 above)
Comment (jbw-odoo) [reply addons/sale/views/sale_order_views.xml:238]: As draft down payments are taken into account in the calculation of the amount when generating final invoice, the idea here is to draw the user attention to the fact that some invoice still are draft.

Comment (Feyensv) [reply addons/sale/views/sale_order_views.xml:238]: I'd prefer some kind of warning in the invoice wizard instead of a duplicated statbutton (and yet another field on sales orders).
The problem with 'nearly duplicated' view elements is that we sometimes get user tickets saying 'oh, i customized this with studio, but in this situation, it doesn't work as expected' because the element they modified or hooked their customization on is not always the one shown.

File: addons/sale/models/sale_order.py (hunk 3 above)
Comment (Feyensv) [reply addons/sale/models/sale_order.py:996]: Well, imo, the down payment invoice(s) are the first ones created, and it's not a problem to clearly see all the existing invoices :thinking:.
Did the PO ask for this change ?

File: addons/sale/views/sale_order_views.xml (hunk 4 above)
Comment (jbw-odoo) [reply addons/sale/views/sale_order_views.xml:238]: > I'd prefer some kind of warning in the invoice wizard instead of a duplicated statbutton (and yet another field on sales orders). The problem with 'nearly duplicated' view elements is that we sometimes get user tickets saying 'oh, i customized this with studio, but in this situation, it doesn't work as expected' because the element they modified or hooked their customization on is not always the one shown.

Ok, make sense.

File: addons/sale/models/sale_order.py (hunk 3 above)
Comment (jbw-odoo) [reply addons/sale/models/sale_order.py:996]: No it's an initiative. The idea is, in the case their already exist one down payment, and you go through the wizard to create a second down payment, if found i more user friendly to display the newly created down payment in form view instead of displaying a list, where the user then must click on the new down payment to view it.

File: addons/sale/views/sale_order_views.xml (hunk 4 above)
Comment (jbw-odoo) [reply addons/sale/views/sale_order_views.xml:238]: @tsb-odoo

**Review by smetl — APPROVED** [review]

**Review by Feyensv — APPROVED** [review]
Minor guidelines/linting remarks.

Also, one conceptual question (also asked to LAS), do we really want multiple downpayment lines on the SO (by tax) or could we simply have one down-payment SOL linked to n AML?  It would give the expected details on the accounting side, while keeping it 'simple' in the sale order :thinking:.

**Comment by smetl** [comment]
@jbw-odoo ci/style red

**Comment by jbw-odoo** [comment]
> @jbw-odoo ci/style red

fixed

**Comment by smetl** [comment]
@robodoo r+ rebase-ff
